source/agora/common/BitMask.d

```diff
@@ -0,0 +1,189 @@
+/*******************************************************************************
+
+    This is an implementation of a bitmask that once initialized will have a
+    fixed number of bits for holding `true` / `false` (`1` / `0`) values.
+    It allocates the required number of `ubytes` in the constructor and does not
+    allow reading or writing to bits beyond the fixed count which is set during
+    construction.
+    This type is created for use as the validators signing bitmask and any
+    changes should ensure that it does not compromise that use.
+
+*******************************************************************************/
+
+module agora.common.BitMask;
+
+public struct BitMask
+{
+    import std.algorithm;
+    import std.conv;
+    import std.format;
+    import std.range;
+
+    @safe:
+
+    /// Serialization
+    public void toString (scope void delegate (scope const char[]) @safe sink) const
+    {
+        formattedWrite(sink, "%s", iota(this.length).map!(i => this[i] ? "1" : "0").join(""));
+    }
+
+    /// Support for Vibe.d serialization
+    public string toString () const
+    {
+        string ret;
+        scope void delegate (scope const char[]) @safe sink = (scope v) { ret ~= v; };
+        this.toString(sink);
+        return ret;
+    }
+
+    pure:
+
+    /// Support for Vibe.d deserialization
+    public static BitMask fromString (scope const(char)[] str)
+    {
+        auto bitmask = BitMask(str.length);
+        str.enumerate.each!((i, c)
+        {
+            if (c == '1')
+                bitmask[i] = true;
+        });
+        return bitmask;
+    }
+
+    nothrow:
+
+    /// Count of active validators who are expected / allowed to sign the block
+    public size_t length;
+
+    /// Bytes to hold bits to indicate if a validator has signed
+    private ubyte[] bytes;
+
+    public this (size_t length) inout
+    {
+        this.length = length;
+        if (length > 0)
+            this.bytes = new inout(ubyte)[1 + ((length - 1) / 8)];
+    }
+
+    public this (size_t length, in ubyte[] bytes)
+    {
+        this(length);
+        assert(this.bytes.length == bytes.length);
+        bytes.enumerate.each!((i, b) => this.bytes[i] = b);
+    }
+
+    // set the bits that are set in given BitMask
+    public auto opOpAssign (string op : "|") (in BitMask add)
+    {
+        assert(this.length == add.length, "BitMask assignment must be for same bit length");
+        iota(this.length).each!((size_t i)
+        {
+            if (add[i])
+                this[i] = true;
+        });
```

Comment:
```suggestion
        this.bytes |= add.bytes;
```

Should work.

Reply:
Note that you can do the same with `&=` if you ever want to have the ability to merge two sets.

Reply:
Updated but needed to `slice` the arrays.